server/models/Document.js

```diff
@@ -223,23 +228,51 @@ Document.searchForUser = async (
 
 // Hooks
 
-Document.addHook('afterCreate', model =>
-  events.add({ name: 'documents.create', model })
-);
+Document.addHook('beforeSave', async model => {
+  if (!model.publishedAt) return;
+
+  const collection = await Collection.findById(model.atlasId);
+  if (collection.type !== 'atlas') return;
+
+  await collection.updateDocument(model);
+  model.collection = collection;
+});
+
+Document.addHook('afterCreate', async model => {
+  if (!model.publishedAt) return;
+
+  const collection = await Collection.findById(model.atlasId);
+  if (collection.type !== 'atlas') return;
+
+  await collection.addDocumentToStructure(model);
+  model.collection = collection;
+
+  events.add({ name: 'documents.create', model });
+  return model;
+});
 
 Document.addHook('afterDestroy', model =>
   events.add({ name: 'documents.delete', model })
 );
 
-Document.addHook('afterUpdate', model => {
-  if (!model.previous('publishedAt') && model.publishedAt) {
-    events.add({ name: 'documents.publish', model });
-  }
-  events.add({ name: 'documents.update', model });
-});
-
 // Instance methods
 
+Document.prototype.publish = async function() {
```

Comment:
Pretty happy with this approach, although it took a while to get it working how I wanted. Basically you can now just deal with document models and add / update properties at will – the hooks will take care of ensuring that the associated collection is updated correctly.

Reply:
this is really nice